sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java

```diff
@@ -0,0 +1,535 @@
+      MongoClient mongoClient = new MongoClient();
+      MongoDatabase mongoDatabase = mongoClient.getDatabase(database);
+      Long collectionEstimatedSize = getEstimatedSizeBytes(options);
+      ArrayList splitIDFromMongo;
+      double initialRatio = 1.6;
+      if (numberSplit > 0) {
+        numberSplitsCalculated = new Long(numberSplit);
+      } else if (desiredBundleSizeBytes > 0) {
+        numberSplitsCalculated = collectionEstimatedSize / desiredBundleSizeBytes;
+        if (numberSplitsCalculated <= 0) {
+          numberSplitsCalculated = 1L;
+        }
+      }
+      if (numberSplitsCalculated == 1) {
+        listEntireCollection.add(this);
+        return listEntireCollection;
+      }
+      int maxChunkSize = 64000;
+      // get the key ranges with splitVector
+      BasicDBObject split = new BasicDBObject();
+      split.append("splitVector", database + "." + collection);
+      BasicDBObject keyPatternValue = new BasicDBObject();
+      keyPatternValue.append("_id", 1);
+      split.append("keyPattern", keyPatternValue);
+      split.append("force", false);
+      split.append("maxChunkSizeBytes", maxChunkSize);
+      Document data = mongoDatabase.runCommand(split);
+      splitIDFromMongo = (ArrayList) data.get("splitKeys");
+      // ratio between number of splits from mongo and splits defined by runner or user
+      double calculatedRatio = (double) splitIDFromMongo.size() / numberSplitsCalculated;
+      if (((calculatedRatio < initialRatio)
+          && (splitIDFromMongo.size() > numberSplitsCalculated))) {
+        ArrayList splitIDCalculated = new ArrayList();
+        double mod = splitIDFromMongo.size() % numberSplitsCalculated;
+        HashMap<Integer, Integer> idRanges = new HashMap<Integer, Integer>();
+        // get the ranges
+        for (int i = 0; i < numberSplitsCalculated; i++) {
+          if (mod >= 0.0) {
+            idRanges.put(i, (int) Math.ceil(calculatedRatio));
+          } else {
+            idRanges.put(i, (int) Math.round((calculatedRatio * 100) / 100));
+          }
+          mod--;
+        }
+        // compute the ranges from the split Vector command
+        Object splitId = 0L;
+        int indexPosition = 0;
+        // isPreviousTwo is used to managed the change of value for the range (from 2 to 1)
+        boolean isPreviousTwo = true;
+        for (Integer range : idRanges.values()) {
+          if (range == 2) {
+            splitId = splitIDFromMongo.get(indexPosition + 1);
+          } else {
+            if (isPreviousTwo) {
+              indexPosition++;
+              isPreviousTwo = false;
+            }
+            splitId = splitIDFromMongo.get(indexPosition);
+          }
+          indexPosition++;
+          splitIDCalculated.add(splitId);
+        }
+        return createSourceList(splitIDCalculated);
+      }
+      return createSourceList(splitIDFromMongo);
+    }
+
+    private List<BoundedSource> createSourceList(ArrayList splitIDList) {
```

Comment:
Seems like it should be possible to make this function static and have its own unit tests.